inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -42,572 +42,547 @@
+}
+namespace {
 
-        if (std::is_same<T, uint8_t>::value && chs == 4) {
-            avx::splitRow_8UC4(in, outs[0], outs[1], outs[2], outs[3], length);
-            return;
-        }
+struct cv_type_id {
+    template <typename type>
+    const int operator()(type_to_type<type> ) { return cv_type_to_depth<type>::depth;}
+};
 
-        if (std::is_same<T, float>::value && chs == 2) {
-            avx::splitRow_32FC2(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                length);
-            return;
-        }
+}  // namespace
 
-        if (std::is_same<T, float>::value && chs == 3) {
-            avx::splitRow_32FC3(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                reinterpret_cast<float*>(outs[2]),
-                                length);
-            return;
-        }
+template <typename typelist>
+bool is_cv_type_in_list(const int type_id) {
+    return type_dispatch<typelist>(type_id, cv_type_id{}, [](...){ return true;}, false);
+}
 
-        if (std::is_same<T, float>::value && chs == 4) {
-            avx::splitRow_32FC4(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                reinterpret_cast<float*>(outs[2]),
-                                reinterpret_cast<float*>(outs[3]),
-                                length);
-            return;
-        }
-    }
-#endif  // HAVE_AVX2
+namespace {
+using merge_supported_types = typelist<uint8_t, int8_t, uint16_t, int16_t, int32_t, float, fp_16_t>;
 
-#ifdef HAVE_SSE
-    if (with_cpu_x86_sse42()) {
-        if (std::is_same<T, uint8_t>::value && chs == 2) {
-            splitRow_8UC2(in, outs[0], outs[1], length);
-            return;
+template<typename T, int chs>
+void mergeRowImpl(scalar_tag, const std::array<const T*, chs>& ins, T* out, const int length) {
+    for (int x = 0; x < length; ++x) {
+        for (int c = 0; c < chs; ++c) {
+            out[chs * x + c] = ins[c][x];
         }
+    }
+}
 
-        if (std::is_same<T, uint8_t>::value && chs == 3) {
-            splitRow_8UC3(in, outs[0], outs[1], outs[2], length);
-            return;
-        }
+template<typename isa_tag_t, int chs>
+struct typed_merge_row {
+    using p_f = void (*)(const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length);
+
+    template <typename type, typename tag = isa_tag_t>
+    typename std::enable_if<std::is_same<tag, scalar_tag>::value ||
+            (!std::is_same<tag, scalar_tag>::value && !std::is_same<type, uint8_t>::value &&
```

Comment:
@sivanov-work please take a look

Reply:
according to previous comment
`tag` is always `isa_tag_t`

so we need to check the second part of SFINAE:

`!std::is_same<tag, scalar_tag>::value && !std::is_same<type, uint8_t>::value && !std::is_same<type, float>::value`
it is ` true && is_NOT__uint8  && is_NOT_float`

So, It fails for another types from list